docs/src/design/decoder/main.md

```diff
@@ -11,7 +11,7 @@ The sections below describe how Miden VM decoder works. Throughout these section
 
 1. An opcode requires $7$ bits to represent.
 2. An immediate value requires one full field element to represent.
-3. A `NOOP` operation has a numeric value of $0$, and thus, can be encoded as seven zeros. Executing a `NOOP` operation does not change the state of the VM, but it does advance operation counter, and may affect program hash.
```

Comment:
This seems incorrect to me:

1. Executing a noop modifies the `clk`, so it does affect the VM's state.
2. The hash of code blocks include the whole content. Good hash functions will avalanche a single bit change, so the presence of a `noop` on any span change's the span contents, which bubbles up the tree until the program's hash changes.

Reply:
I agree with the first point, and we should probably modify the statement to be something like: "Executing a `NOOP` operation does not change state of the VM except for the operation counter, which is incremented by one."

Regarding the second point: executing a `NOOP` *may* affect program hash but that's not always the case. Whether it does or does not, depends on the exact position of the `NOOP` within operation batch and group. For example, a group containing no operations (i.e., with numeric value of `ZERO`) can be executed as a single `NOOP` or as 9 `NOOP`'s. Regardless of how many `NOOP`'s we execute, we still hash the same value of `ZERO` - so, there is no change in the input into the hash function.

Reply:
It seems I missed the point this is trying to make. Why is the execution of the `NOP`s something that is being emphasized?

No other operation had this emphasis (well, at least on the decoder docs), and AFAIU the execution of the operations doesn't affect the program hash, only their contents, no? (By program hash, I understand the MAST).

Reply:
`NOOP` is the only operation execution of which may not affect program hash. Basically, if we execute any other operation an additional time, the program hash (MAST root) is guaranteed to change. But in some situations, we can execute an additional `NOOP` and program hash will remain the same.

Reply:
I have two interpretations of what you're saying:

1. Adding a new `NOOP` to the instruction stream, and executing it doesn't change the program hash.
2. The VM may not execute the existing `NOOP`s in the instruction stream, and that doesn't affect the program hash.

For case `1`, I don't understand how that would be true, adding instructions would change the hash input, and change the program hash, no?

For case `2`, then I get it, and it is a surprising fact to me that the VM may not execute all the existing `NOOP`s, if that is the case, may I add some additional mention to that?

Reply:
It is both, actually. Basically, there are situations where the following programs would have the same MAST root:
```
<some instructions> NOOP NOOP <some more instructions>
```
```
<some instructions> NOOP <some more instructions>
```
```
<some instructions> NOOP NOOP NOOP <some more instructions>
```
For such a program, the VM could execute either 1, 2, or 3 `NOOP` in this specific location and it wouldn't affect program hash computation.

The underlying reason for this is that the numeric value of `NOOP` is 0 and that a single field element can encode multiple operations. Basically, adding adding many zeros to a zero is still zero - so, one `NOOP` is sometimes indistinguishable from 2 `NOOP` etc.

Reply:
I can see how your argument applies to MAST assembly, something like:

```
let b = SpanBuilder::new(None);
b.add_op(Operation::Add);
let blocks = vec![];
b.extract_final_span_into(blocks);
```

In the snippet about there could be up to 55 explicit NOOPs added before the extract, and the hash would not change. That is the case because these 55 NOOPs are implicitly added anyways, and the `add_op(Operation::Noop)` is effectively a no-op (pun intended :) ).

But I don't understand how that plays a whole during decoding of a running program. The number of padding `NOOP`s was chosen by the assembler, and the program hash commits to that choice, it seems weird that adding or removing opcodes still result in the same hash.

Am I misunderstanding the point here?

Reply:
This doesn't have much to do with the assembler but more with how operations are arranged in a `SPAN` block ([see here](https://github.com/0xPolygonMiden/miden-vm/blob/main/core/src/program/blocks/span_block.rs#L19)). Basically, there is a concept of an *operation group* which is a single field element encoding up to 9 operations. If a value of an operation group is `ZERO` it could encode any number of `NOOP`'s between 1 and 9 (this is because value of `NOOP` operation is `ZERO`) and so, the string that we hash to compute a MAST root would look exactly same for 1 or 9 `NOOP`s in a group.